README
FFP 3.18 - Feature Frequency Profile Phylogenetics Package
Feb 20, 2012

Author: Gregory E. Sims


This is a collection of programs / utilities for implementing
the FFP (Feature Frequency Profile) method of phylogenetic
comparison.  FFP is a class of alignment-free methods suitable
for (whole genome) comparisons from viral to mammalian scale 
genomes.

This method has been used to perform various phylogenetic
analyses:


Sims GE and Kim SH (2011)  Whole-genome phylogeny of Escherichia 
coli/Shigella group by feature frequency profiles (FFPs). PNAS, 108, 
8329-34.

Jun SR, Sims GE, Wu GA, Kim SH. (2010) Whole-proteome phylogeny of
prokaryotes by feature frequency profiles: An alignment-free method 
with optimal feature resolution. PNAS, 107,133-8.

Sims GE, Jun SR, Wu GA, Kim SH. (2009) Alignment-free genome comparison 
with feature frequency profiles (FFP) and optimal resolutions.
PNAS, 106,2677-82.

Sims GE, Jun SR, Wu GA, Kim SH (2009) Whole-genome phylogeny 
of mammals: evolutionary information in genic and nongenic regions. 
PNAS. 106,17077-82.



The utilities are designed to be implemented using unix
command pipes.  In other words the output of programs
can be linked to the input of other programs.  Therefore
many of the scripts are acceptable as filters to be used
in intermediate steps.


This package contains the following programs/scripts:


ffpgui		[Experimental] A perl/Tk based GUI 
                interface for performing some of the
                basic FFP operations.  This utility 
                doesn't support grid-based/ 
		multiprocessor job flow.  Also ffpgui is
		in the beta stage, but it should give an
		example of what can be done with the 
		utilities listed below. Note, currently 
		ffpgui 	will only work properly in Cygwin 
		if you 	download and compile v804.029  
		perl/Tk module from CPAN. Automatic installation 
		of perl/Tk by Cygwin setup.exe will not
		provide proper functionality (It uses a
		much older version version of perl/Tk).
		See INSTALL for further instructions.
		Use --disable-gui to bypass installation. 

ffpry		Constructs an FFP profile from nucleic acid 
		sequences in FASTA format (.fna). 

ffpaa           Constructs an FFP profile from amino acid 
		sequences in FASTA format (.faa).

ffprwn		This performs row normalization of the raw
		FFP matrix.

ffpjsd		This calculates the Jensen Shannon Divergence
		between FFPs and outputs a Divergence
		(Distance) matrix.  A variety of other 
                distances/similarity metrics are available as
		well.

ffpboot		This performs bootstrapping or jacknifing
		permutation of a raw FFP profile produced
		by ffpry or ffpaa

ffpvocab	This utility counts the number of words which
		are used more than a paritcular threshold
		in the FFP profile. This utility is used
		to determine what is the best range of word
		lengths to use for a genome collection

ffpre		This utility calculates the Relative entropy
		between the expected and observed frequencies
		of features of length l (specified on the 
		command line) using an L-2 Markov Model.

ffpvprof	Script which calculates the word usage 
		for a range of l.  Runs ffpvocab

ffpreprof	Script which calculates the Relative entropy
		between observed and predicted frequencies for
		a range of l.  Runs ffpre.

ffpmerge	This utility merges all rows of an FFP into
		a single row.  Use this for merging segments
		of an FFP, for example different chromosomes
		of a larger genome.

ffpcol          This utility converts a FFP which has been
		written out in key/value format to a columnar
		format, so that each column corresponds to
		the same feature in each row of the FFP.           

ffptxt		This utility creates a key/value FFP of text
                data.  This is useful for performing an FFP
		analysis of human language texts.  All non-
		alphanumeric characters are ignored. 

ffpfilt		Eliminate high/low frequency features using 
                frequency cutoffs or probability based cutoffs 
                assuming a normal or extreme value distributions.

ffpcomplex      Eliminate high/low complexity features using
                a complexity cutoff or probability based cutoff 
                assuming a normal distribution.

ffpdf		Finds clade distinguishing (diagnostic) features.
		See Sims GE and Kim SH (2011) PNAS 108.

ffptree		Build neighbor joining and UPGMA trees from ffpjsd
		output.


OTHER REQUISITE PROGRAMS

We suggest that you obtain a copy of PHYLIP 
(http://evolution.genetics.washington.edu/phylip.html)
for building trees, however you can use any tree building program
which will accept distance matrix input.   The utility ffpjsd
will produce Phylip style 'infile's as well as raw distance
matrices.  As of version 3.06, a tree building utility, ffptree
is included, which will allow you build Newick style tree output
directly as part of a ffp pipeline, which is compatible with the
Phylip (3.69) utilities.


QUICK START

The best way to get a quick start is to read through the 
simple tutorial PDF file located in the ./doc dir of this
distribution.  It is also installed during the make install
process in the /usr/local/share/doc/ffp directory (unless
you have specified a different base directory using
./configure --prefix during the build process).


EXAMPLES

***** How do I perform FFP comparison on a collection of nucleic
      acid sequences, using a particular length of feature?


Assuming your nucleic acid .fna files are all in the current
working directory and are named with the .fna extension:

ffpry -l 5  *.fna | ffpcol | ffprwn | ffpjsd > matrix

for just two files (test1.fna and test2.fna):

ffpry -l 5  test1.fna test2.fna | ffpcol  | ffprwn | ffpjsd > matrix


The above example uses all features of length
5.  The output of ffpry will be in key-value form,
i.e. pairs of feature sequence followed by the raw
count.  Each row corresponds to the features from
that sequence, in the order of input.The output of 
ffpry is piped to ffpcol, which converts the key value
form into a column form, so that the raw counts 
corresponds to the same feature across rows. The 
utility ffprwn row normalizes each  row of the ffp
feature matrix (output by ffpry), so that each 
element of that row is a relative frequency. 
The output is now piped to ffpjsd which calculates a 
Jensen Shannon Divergence Matrix.

Alternatively you can save the output at each step in 
intermediate files in the following form:

ffpry -l 5 *.fna > vectors
ffpcol vectors > vectors.col
ffprwn vectors.col > vectors.row
ffpjsd vectors.row > matrix

This may be useful if you want to perform multiple analyses
on some intermediate file (For example bootstrapping -- see
below).



**** How do I script and run commands?


All of these commands can be completed 
programmatically in a shell script file 
for example:

In a file named, for instance ffptest.sh


#!/bin/sh

ffpry -l 5 *.fna > vectors
ffpcol vectors > vectors.col
ffprwn vectors.col > vectors.row
ffpjsd vectors.row > matrix

Save the script and make it executable using:

chmod +x ffptest.sh

then run the script from the command line

./ffptest.sh


**** How do I perform bootstraping?

You can use the utility ffpboot to perform bootstrapping on
the output of ffpry or ffpaa.


ffpry -l 5 *.fna | ffpcol > vectors
ffpboot vectors | ffprwn | ffpjsd > matrix



**** How do I create multiple bootsrap sets?

The example below creates 100 bootstrap pseudoreplicate
JSD matrices.

ffpry -l 5 *.fna | ffpcol > vectors
for i in $(seq 1 1 20)
	do
	ffpboot vectors | ffprwn | ffpjsd > matrix.$i
	done


**** How do I create phylip format infiles?


To create phylip format infiles to use with programs
such as NEIGHBOR, Use the command ffpjsd -p [FILE] which 
will generate a phylip format infile. FILE specifies
the names of the taxa in the fna files you are using

This file should whitespace or newline delimit the
different taxa names.


i.e.

Taxa_1
Taxa_2
...
Taxa_N


Note the taxa should be in the order that they
are read into ffpry (use ls *.fna to get that
ordering).

For example:

ffpry -l 5 *.fna | ffpcol | ffprwn | ffpjsd -p names.txt > infile

Or without the wildcards (*.fna)

ffpry -l 5 1.fna 2.fna 3.fna | ffpcol | ffprwn | ffpjsd -p names.txt > infile

The file names.txt should contain the taxa names of 1.fna, 2.fna and 3.fna
in that order.

**** How do I build a neighbor joining tree? ****

In the sprit of the pipeline concept you can pipe output directly from
ffpjsd into ffptree, provide it is in phylip infile format. Continuing
the example from above:

ffpry -l 5  *.fna | ffpcol | ffprwn | ffpjsd -p names.txt | ffptree -q > tree

This produces a tree in Newick format.  If you want to see the
human readable tree to, remove the -q switch.  Note, lots of 
output will be produces, but written to standard error so you will
need should redirect standard error to a file to save for later.

ffpry -l 5 *.fna | ffpcol | ffprwn | ffpjsd -p names.txt | ffptree  2> progress > tree


**** How do I do bootstrapping for use with Phylip?


ffpry -l 5 *.fna | ffpcol > vectors
for i in $(seq 1 1 20)
        do
        ffpboot vectors | ffprwn | ffpjsd -p species.txt >> infile
        done

This will create a multiple dataset file for use with phylip.
Use the 'multiple datasets' option in the neighbor program.


**** How do I perform FFP on large genomes?


The most effective way to do this to calculate FFP's of 
segments or units of the genome, for instance by chromosome
or by contig, the ffp's of individual units can be
merged together using the ffpmerge utility.

Say you have 10 chromosomes. Calculate the FFP of
each as a separate process and merge at the end of
calculations.

This is especially effective for multiprocess machines.


For example:

for fna_file in $(ls *.fna)
	do
	ffpry -l 10 $fna_file > $fna_file.vector &
	done

ffpmerge *.vector > merged.vector


If your cluster machine uses a qeueing system (i.e. Grid 
engine) then you can create individual shell scripts to
give to the scheduler and then merge unit vector files after
all scheduled jobs have completed.  A simple example using
grid engine employs the $SGE_TASK_ID variable.

Save a file containing the paths to your sequences
There are 10 files total
$ cat > sequences.txt
seq.fna
seq2.fna
seq3.fna
....
Ctrl-D

$ cat > submit.sh
#!/bin/bash
#submit.sh

FILE=`head -n $SGE_TASK_ID <  $1 | tail -n 1`
ffpry -l 10 $FILE   > $FILE.vector &

In your shell:
chmod +x submit.sh

qsub -a 1-10 submit.sh sequences.txt

Ctrl-d

**** What if the ffp I get from ffprwn is very large and
     ffpjsd take a long time?

In this case you may want to break up the calculation of
the JSD matrix, by assigning specific rows to different
CPUs using the -r option.

Starting with a normalized ffp with 10 rows:

ffpjsd -r 1 vector.row > row.1

The other 9 rows can be calculated by other CPUs, once 
again using a cluster machine and the SGE_TASK_ID variable. 
The results can be merged again using shell scripting:


for i in $(seq 1 1 10)
	do
	cat row.$i >> matrix
	done



**** What is the difference between key/value FFPs and 
columnar FFPs?

A key/value FFP is an FFP form which is generated by
default from the programs FFPry and FFPaa.  For instance
the following command will generate a FFP of this form:

ffpry -l 5 test*.fna


The format will resemble:

RYRRR 2 RRRRY 3 RRRRR 0 ....
YRRRR 1 YYYRR 1 YRRRY 2 ...
...


Each row of the file is a FFP derived from a different
sequence file.

Columnar formats are required for input to ffprwn and
ffpboot.


In this format no feature keys are printed and the columns
in the file correspond to the counts of that feature in
each of the sequence files.  For very sparse FFPs the key/
value FFP can generate smaller files.  For example this 
command will generate a key-value FFP:

ffpry -l 5 test*.fna


To convert a key/value FFP into a columnar format for input
to the other utilities the ffpcol utility should be used
as a filter.

ffpry -l 5 test*.fna | ffpcol


The output from ffpcol can be used in ffprwn and ffpjsd


ffpry -l 5 test*.fna | ffpcol | ffprwn | ffpjsd 


**** How do I use a full 4 letter Nucleotide or 20 letter 
     amino acid alphabet?

By default character classing is used in both ffpry and 
for amino acids in ffpaa. To disable this classing specify
option -d for ffpry, ffpaa and ffpcol.  Please also take
note that when you disable RY coding with the -d option
you may need to add the -d option to subsequent filters
such as ffpcol and ffpmerge.

For example:

ffpry -l 5 -d test*.fna | ffpcol -d | ffprwn | ffpjsd

For amino acids:

ffpaa -l 5 -d test*.faa | ffpcol -d -a | ffprwn | ffpjsd

**** How do I use a spaced seed hash with FFP?

FFP refers to spaced seeds as masks - from the
manner in which masks are used in computer programming
to 'mask out' certain bit positions in low level bit
manipulations of numbers stored in binary format.
A spaced seed or mask of '01110' will allow both
CAAAG and TAAAA to match each other, as well as to
match any 5 letter word with AAA in the middle.

A mask can be specified using -w, for example:

ffpaa -l 5 -d -w "01110" test*.faa | ffpcol -d -a | ffprwn | ffpjsd

If you don't want explicitly supply a mask string
but want to allow a certain number of mismatches, use
-z.  Here for example is how to create a random mask
with two mismatches allowed.

ffpaa -l 5 -d -z 2 test*.faa | ffpcol -d -a | ffprwn | ffpjsd

**** How do I compare text files with FFP?

FFP has the ability to compare text files with the utility
ffptxt.  The procedure is much the same as with nucleic acid
and amino acid sequences.  Specify text with the -t option
when using ffpcol.

ffptxt -l 4 file*.txt | ffpcol -t | ffprwn | ffpjsd

**** How do I get help?

All the utilities come with their own manual page which is installed
by default when using 'make install'.

man ffpry

will retrieve the manual for the ffpry utility.

If you have installed ffp in some alternate location (i.e.
by using ./configure --prefix), then the locations of the
ffp manuals may not be in your MANPATH environmental variable.
You can add the manual directory to MANPATH, or simply
read the manuals directly using 

man /pathtomanual/ffpry.1

(Note the manual section extension '1').


**** Example trees

Some published examples are shown in the distribution
'examples' directory.

**** How do I run FFP in Windows?

Use Cygwin (www.cygwin.com).  This package has been developed
and tested to perform in both Linux and Cygwin.
Cygwin is designed as an emulated Linux/Unix 
environment which runs on the Windows operating
system -- it includes the majority of the GNU compilers
and utilities that are part of a standard Linux
distribution and performs superbly (albeit with
a small performace loss because of emulation).


**** Multiple fastas in a single file

By default the ffpry and ffpaa programs assume that a single
file, regardless of the number of fasta records contained in
that file, represents a single species/genome/proteome. Therefore
the l-mer frequencies represent the counts for all fasta records.
If you specify multiple files on the command line i.e.

ffpry *.fna

Which might expand (the expanding of which is done by your shell of
course) to:

ffpry test1.fna test2.fna test3.fna

then 3 separate ffp lines will be printed in the output.  If in fact
you want all of these results to be merged together into one FFP you
can use the ffpmerge utility, or simply use the cat command, both
of which should produce equivalent results.

cat *.fna | ffpry 

or

ffpry *.fna | ffpmerge --keys

If however you have a single (or multiple fasta files) with multiple
records which you want to be individual FFPs then you must specify
the -m option.

ffpry -m *.fna


**** How do I implement the alternate Hamming based distance refered to
as the Evolutionary FFP distance in Sims and Kim (2011), PNAS, 108
8329-34? 

Trees presented in this paper are included in the examples subdirectory.
To implement this type of analysis on your own use the following 
snippet of shell script code (assuming you are using the bash shell).

From your working directory containing all your genome fasta files
(assuming small single fasta genomes).


ffpry -l 20 *.fasta | ffpcol | ffpfilt -l 0.05 -u 0.95 -e > ffp.filtered

for i in {1..100} ; do
      ffpboot -j -p  0.1  ffp.filtered | ffpjsd -H -p species.txt
done > infile


The features are filtered to remove high and low frequency features.
Then 100 pseudoreplicates are created using 10% jackknife sampling.
The output is a PHYLIP style infile which can be used directly as 
input to the PHYLIP tool NEIGHBOR.  The option argument to -p is a
tab or newline delimited file containing the names of the taxa in the
original order which was specified on the command line to the original
ffpry invocation.  To confirm you have taxa named in the right order
in your 'species.txt' taxa name file, execute this shell expansion.

echo *.fasta

This will show you the order (which will be identical to the ordering
observed using ls *.fasta), in which you need to specify the names in
species.txt.

Rather than relying on the order from shell expansion you can specify
the genome file arguments explicitly

ffpry -l 20 genome1.fasta genome2.fasta genome3.fasta ... 

See the examples subdirectory for more information, including sample
trees generated using FFP and Phylip.

**** How do I implement the block-FFP method  mentioned in
Sims GE, Jun SR, Wu GA, Kim SH. (2009) Alignment-free genome comparison
with feature frequency profiles (FFP) and optimal resolutions.
PNAS, 106,2677-82.

There is currently no script included in this distribution which will
implement this method of genome comparison. Future releases will
contain executables which implement Block-FFP.

The main point to keep in mind is that FFP works best when you
are comparing genomes/sequences of similar length -- and a good 
guideline is to make sure that your genomes are within A-fold 
the size of each other where A is the number of symbols in your
alphabet.


********





Copyright (C) 2009-2012
Author: Gregory E. Sims

Report Bugs to gsims1997@yahoo.com




NEWS
Version 3.19
	Thanks to Joemar Taggana, some incompatible utilities 
	were discovered in Mac OSX -- including incompatible
	getopt and mktemp.  Corrected confusion in tutorial 
        p. 3 regarding -m (Thanks to Erin Shellman).  Corrected
	spelling of ffpjsd option --hamann to --hamman.

Version 3.18
	Rearranged ungetc character tests to come after
	conversion of pipe to file.  Fixed memory leak in
	hash init().  Added support for large files (> 2GB) on 32
        bit machines.

Version 3.17
	Fixed bug in tempfile creation in function 
        convertPipeToFile().

Version 3.16
	Fixed bug in ffpry with multiple headers parsed as a
	single genome entry.

Version 3.15
	Fixed bug in ffprwn.

Version 3.14
	Added --disable-gui options to ./configure script. This
	disables the perl based gui, so that the Perl/Tk module
	isn't required for installation.

Version 3.13
	Changed row terminal output for all programs from '\t\n' 
	(tab newline) to '\n' (newline) only.  This makes output
	from previous version incompatible with 3.13 but the
	ffps can be easily converted using:

	sed 's/\t$//' ffp > ffp.tmp ; mv ffp.tmp ffp

Version 3.12
	Small bug fix for ffprwn for GNU compiler on 64 bit machines
	where promotion of unsigned int to long unsigned int is
	different.  Thanks to Adrienne Breland at the University
	of Nevada, Reno for finding this.

Version 3.10
	Fixed minor bug in ffpry for fasta records pseudomolecule 
	input ENDING in NNNNNNNN, or other invalid characters.

Version 3.09
	Fixed bug in ffpdf, related to initialized GROUP variable.
	
	Updated license

Version 3.08

	ffpgui

	Added instructions to install a functional version of the
	perl/tk module for using cygwin, in INSTALL.

	Added error messages in ffpgui for running a vocab or 
	relative entropy profile without loading a file first.

	Added additional sections to tutorial.pdf
       


Version 3.07
	Added signal handling to all executables.  
	Broken pipes will cause a termination error in a pipeline
	and will output 'PROGRAM: Broken pipe'

	Expanded ./doc/tutorial.in.
	Updated license.
	Fixed inappropriate default -f setting (changed from 1 to
	2) for ffpvprof.

Version 3.06
	Added ffptree utility, which creates neighbor joining and
	upgma trees.  Date and version automatically updated 
	during build process.
	
        Improved seeding of random generator when no seed is
	provide via command line option:  ffpry, ffpboot, ffpaa

	Added processing via ffptree to ffpgui.  The human 
	readable tree is displayed in the message window.

	Started tutorial file in ./doc directory. 

	Added tutorial.pdf to doc directory

Version 3.05
	Added length test since MAX_WORD_SIZE is 40: ffpry, ffpaa
	ffpre.

Version 3.04
	Added script ffpgui which is a Perl/Tk based GUI wrapper of
	the various utilities.

	Added --keys flag to ffpfilt to force keys to be output
	instead of values only.

Version 3.03
	Fixed overflow error with non-rabin karp hash used in ffpcol.

Version 3.02
	The ffpaat utility has been removed from this package.  The
	purpose of this utility was to count amino acid features in 
	all nine frames from DNA fasta input.  However	obtaining
	a proper translation of gene regions and using ffpaa is 
	generally more informative.  This utility can still be 
	compiled and used from previous versions, but will not longer
	be part of future releases. 

	Updated ffpre to Rabin-Karp hash, also fixed float overflow
	errors and cases of nan, or inf in output.  Text mode is now
	implemented in ffpre.

	Updated ffptxt to Rabin-Karp hash.


Version 3.01
	Implementation of Rabin-Karp hash in ffpaa utility.
	Similar speed gains as in ffpry obtained


	ffpaa -l 20 NC_008253.faa      

	ffpaa (v.3.0)                        ----------- 9.35 s 
	ffpaa (v.3.01)                       ----------- 0.15 s

	Input buffering options removed from both ffpaa and ffpry. The
	optimal buffering size for the system is determined automatically
	by using the stat.st_block size attribute of fstat or the compiler
	default, whichever is greater.  This is an effort to use the
	filesystem's most efficient block size.  
	
	Upcoming v3.02 will include Rabin-Karp implementation of ffpre.

Version 3.0
	Thanks to Nick Loman of University of Birmingham for pointing out
	the cryptic nature of the "key-value" error messages when both
	ffpfilt and ffpcol are used together in a pipeline successively.
	
	All error messages from the utilities state the program issuing
	the error.

	Revamped tmpfile creation for non-seekable piped input, to use
	most efficient buffering.

	Added extra unit tests to confirm piped vs file argument output
	is identical.

	ffpry changes:

	Beginning initial development of Rabin Karp Rolling hash function
	implementations which has replaced the hashing function used
	in version 2 in the ffpry executable.  This has provided a 
	significant speedup in k-mer counting.
	Here is a speed comparison:
	
	ffpry -l 10 -d NC_008253.fna             ------    9.7s (v.3.0)
        ffpry -l 10 -d NC_008253.fna             ------ 1m28.2s (v.2.30

	Removed all-possible/-a flag from ffpry.  The only way to generate
	columnar output from ffpry is via the -f flag and specify a specific
	feature set.

        -f specified features are now reverse complemented and classed as
	is the input sequence by default.	

	Added -r option which disables reverse complementing of the input
	sequence as well as operates on the feature list.  
	
	Fixed ffpry -f long option to correct value as per man file.
	
	Other:

	Fixed segmentation error in ffpboot

	Future directions:

	Rabin-Karp hash function will be implemented in ffpaa, ffptxt and
	ffpre, ffpaat
	
	

Version 2.30
	Fixed bug when encountering IUPAC codes in input files to ffpry.

Version 2.29
	Fixed bug when piping input into ffpmerge.

Version 2.28
	Addition of examples directory with trees from 
	Sims & Kim(2011) PNAS,108,8329-8334

        BUGS: ffpre doesn't report errors for trying to open

src/ffpry.c
/*****************************************************
* This code is distributed under a Non-commercial use 
* license.  For details see LICENSE.  Use of this
* code must be properly attributed to its author
* Gregory E. Sims provided that its use or derivative 
* use is non-commercial in nature.  Proper attribution        
* can be made by citing:
*
* Sims GE, et al (2009) Alignment-free genome 
* comparison with feature frequency profiles (FFP) and 
* optimal resolutions. Proc. Natl. Acad. Sci. USA.
* 106, 2677-82.
*
* Gregory E. Sims (C) 2010-2012
*
*****************************************************/

/* Current revision: $Revision: 1.37 $ 
 * On Tag name:  $Name:  $ 
 * Date commited: $Date: 2012-02-26 01:57:45 $
 * Author: $Author: gsims $ 
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <getopt.h>
#include <math.h>
#include <ctype.h>
#include <errno.h>
#include <unistd.h>
#include <sys/stat.h>
#include "hashroll.h"
#include "mask.h"
#include "ffpry.h"
#include "utils.h"
#include "vstring.h"
#include "sighandle.h"
#include "parse_features.h"
#include "../config.h"


#define DEFAULT_WORD_LENGTH 10  /**< Default Feature length if not given by opt -l */
#define CHAR_BUFFER_SIZE 10000	/**< Default buffer size for reading sequence files */


/* Function prototypes */

static void parseFile(HASH *, FILE *);
static void loopFeatureList(HASH * h, FILE * fp);
static void loopRaw(HASH * h, FILE * fp);


/* Global variables */

char PROG_NAME[FILENAME_MAX];
char *weightVector;               /**< A mask to allow mismatches in features */
int Length = DEFAULT_WORD_LENGTH; /**< Feature length to use if not specified by opt -l */
int maxWordSize = MAX_WORD_SIZE;  /**< Maximum allowed length for opt -l */
long Buffsize = CHAR_BUFFER_SIZE; /**< File input buffer, increase value for larger genomes. opt -b can be used to change this value */


/* Option flags and option arguments */
char *cvalue = NULL;    /**< Option argument string specifying character mask, should change to w */
char *fvalue = NULL;    /**< Option argument file name specifying a file of containing features */
bool zflag = false;	/**< Create a random character mask, -z */
int zflagN = 0;	        /**< Option argument to -z */
bool sflag = false;	/**< Specify a random seed, -s */
int sflagN = 0;	        /**< Option argument to -s */
bool qflag = false;	/**< Use quiet option */
bool wflag = false;	/**< Use a character mask, -w */
bool fflag = false;	/**< Use a list of features specified in a file, -f */
bool dflag = false;	/**< Disable RY coding, -d */
bool mflag = false;	/**< Multiple sequences in one file, -m */
bool rflag = true;	/**< Do reverse complement */


char usage_str[] = "Usage: %s [OPTIONS]... [FILE]... \n\
This program generates an FFP vector of RY-coded features\n\n\
Given no options the default behavior of the program is to\n\
generate a Feature Frequency Profile (FFP) using features of length\n\
10.\n\
\t-l LEN, --length=LEN\n\
\t-f FILE, --feature-list=FILE\n\
\t-w STR, --mask=STR\n\
\t-z K, --random-mask=K\n\
\t-s INT, --rand-seed=INT\n\
\t-q, --quiet\n\
\t-d, --disable\n\
\t-r, --disable-rev\n\
\t-m, --multiple\n\n\
Copyright (c) %s\n\
%s\n\
Contact %s\n";



int main(int argc, char **argv)
{
    int opt;
    FILE *fp;
    HASH h;
    int option_index = 0;

    static struct option long_options[] = {
	{"length", required_argument, 0, 'l'},
	{"mask", required_argument, 0, 'w'},
	{"disable", no_argument, 0, 'd'},
	{"random-mask", required_argument, 0, 'z'},
	{"quiet", no_argument, 0, 'q'},
	{"random-seed", required_argument, 0, 's'},
	{"help", no_argument, 0, 'h'},
	{"feature-list", required_argument, 0, 'f'},
	{"multiple", no_argument, 0, 'm'},
	{"disable-rev", no_argument, 0, 'r'},
	{"version", no_argument, 0, 'v'},
	{0, 0, 0, 0}
    };


  initSignalHandlers();

  strcpy(PROG_NAME,basename( argv[0] ));

    while ((opt = getopt_long(argc, argv, "l:dw:z:s:qf:hmvr",
			      long_options, &option_index)) != -1)
	switch (opt) {
	case 'l':
	    Length = atoi(optarg);
	    break;
	case 'w':
	    if ((weightVector = (char *) malloc(sizeof(char) * strlen(optarg)))==NULL) 
		fatal_at_line("%s\n",strerror(errno));
	    strcpy(weightVector, optarg);
	    wflag = !wflag;
	    break;
	case 'z':
	    zflag = !zflag;
	    zflagN = atoi(optarg);
	    break;
	case 's':
	    sflag = !sflag;
	    sflagN = atoi(optarg);
	    break;
	case 'q':
	    qflag = !qflag;
	    break;
	case 'd':
	    dflag = !dflag;
	    break;
	case 'm':
	    mflag = !mflag;
	    break;
	case 'f':
	    fflag = !fflag;
	    fvalue = optarg;
	    break;
	case 'r':
	    rflag = !rflag;
	    break;
	case 'v':
	    printVersion();
	    exit(EXIT_SUCCESS);
	    break;
	case 'h':
	    printUsageStr();
	    exit(EXIT_SUCCESS);
	    break;
	default:
	    printErrorUsageStr();
	    exit(EXIT_FAILURE);
	    break;
	}

    if ( getenv("MAX_WORD_SIZE") ) 
	maxWordSize=atoi(getenv("MAX_WORD_SIZE"));	    

    if (Length > maxWordSize) 
	    fatal_msg("%d: Max Word size is: %d",Length,maxWordSize);

    if (sflag)
	srand((unsigned) sflagN);
    else
	srand((unsigned) time(NULL) * getpid());


    if (wflag) {
	if (strlen(weightVector) != Length)
	    fatal_msg("length of mask not equal to feature length.\n");
	if (!qflag)
	    warn_msg("Using feature mask: %s\n", weightVector);
    }



    if (zflag) {
	while (zflagN > Length - 1)
	    zflagN--;
	if ((weightVector = (char *) malloc(sizeof(char) * (Length + 1))) == NULL) 
		fatal_at_line("%s\n",strerror(errno));
	randweight(weightVector, Length, zflagN);
	if (!qflag)
	    warn_msg("Using feature mask: %s\n", weightVector);

    }

    //Initialize a hash
    init(&h, (zflag || wflag), nucleotide, !dflag, rflag, Length);

    //Fill with keys if restricting to a feature list
    if (fflag)
        parseFeatureList(&h,fvalue,Length,nucleotide);

    // Must now process file arguments



    argv += optind;

    do {
	fp = stdin;
	if (*argv) {
	    if (!strcmp(*argv, "-"))
		fp = stdin;
	    else if ((fp = fopen(*argv, "r")) == NULL)
	    	fatal_msg("%s: %s\n", *argv, strerror(errno));

	    if ( isDirectory(*argv) ) 
	    	fatal_msg("%s: %s\n",*argv, strerror(EISDIR));

	    argv++;
	} else if (isatty(STDIN_FILENO))
	    printErrorUsageStr();

	parseFile(&h, fp);

	if (fp != stdin)
	    fclose(fp);

    }
    while (*argv);

    //freeHash(&h);

    return EXIT_SUCCESS;
}


/**
 *
 *  Parses FASTA format FNA file
 *
 *  This function can read from stdin or from fasta files.
 *  The hash h, is populated with features differently
 *  if there is a feature list passed via the -f option.
 *
 *
 *  @param h The hash table to use.
 *  @param fp A file pointer to the nucleotide fasta file to parse.
 *  @return none
 *
 */


static void parseFile(HASH * h, FILE * fp)
{
    if (fflag)
	loopFeatureList(h, fp);
    else
	loopRaw(h, fp);

}


/**
 *
 *  Populates a hash with feature counts from a Nucleotide fasta FILE
 *  ptr restricting the search to features defined in the feature list
 *
 *  This function can read from fp which points to stdin or a fasta file.
 *  Function behaves differently if a spaced seed (or mask) is defined with
 *  -z or -w.
 *
 *
 *  @param h The hash table to use.
 *  @param fp A file pointer to the nucleotide fasta file to parse.
 *  @return none
 *
 */

static void loopFeatureList(HASH * h, FILE * fp)
{
    struct stat fattr;
    size_t optimal_size;
    ssize_t nr;
    char *buf;
    bool firstRecord=true;
    
    //Determine optimal buffer size for file
    if (fstat(fileno(fp), &fattr))
	fatal_msg("fd %s: File stat error.\n",fileno(fp));

    //Find optimal size for Disk IO 
    optimal_size = (fattr.st_blksize >= BUFSIZ) ? fattr.st_blksize : BUFSIZ;

    if ((buf = (char *) malloc(optimal_size))==NULL) 
		fatal_at_line("%s\n",strerror(errno));

    // replace w/ function pointers
    while ((nr = fread(buf, sizeof(char), optimal_size, fp)) != -1 && nr != 0) {

	if (zflag || wflag)
	    chkpushatgcw(h, buf, nr,firstRecord);
	else
	    chkpushatgc(h, buf, nr,firstRecord);

	firstRecord=false;
    }
    printFeatures(h);
    free(buf);
    freeHash(h);
}


/**
 *
 *  Populates a hash with feature counts from a Nucleotide fasta FILE
 *  ptr.
 *
 *  This function can read from fp which points to stdin or a fasta file.
 *  Function behaves differently if a spaced seed (or mask) is defined with
 *  -z or -w.
 *
 *  @param h The hash table to use.
 *  @param fp A file pointer to the nucleotide fasta file to parse.
 *  @return none
 *
 */

static void loopRaw(HASH * h, FILE * fp)
{
    struct stat fattr;
    size_t optimal_size;
    ssize_t nr;
    char *buf;
    bool firstRecord=true;
    

    //Determine optimal buffer size for file
    if (fstat(fileno(fp), &fattr))
	fatal_msg("fd %s : File stat error.\n",fileno(fp));

    //Find optimal size for Disk IO 
    optimal_size = (fattr.st_blksize >= BUFSIZ) ? fattr.st_blksize : BUFSIZ;

    if ((buf = (char *) malloc(optimal_size))==NULL) 
		fatal_at_line("%s\n",strerror(errno));

    // replace w/ function pointers
    while ((nr = fread(buf, sizeof(char), optimal_size, fp)) != -1 && nr != 0) {

	if (zflag || wflag)
	    pushatgcw(h, buf, nr,firstRecord);
	else
	    pushatgc(h, buf, nr,firstRecord);

	firstRecord=false;
    	
    }


    printFeatures(h);
    free(buf);
    freeHash(h);
}


